### ISSUE TITLE
`at_pointer` doesn't like an empty string JSON Pointer

### ISSUE THREAD

**Issue by nigeltao** [simdjson/simdjson#1142]
This snippet:

    simdjson::dom::element element = parser.parse(input_array, input_length);
    element = element.at_pointer("");

crashes:

    a.out: simdjson.h:1357: constexpr const CharT& nonstd::sv_lite::basic_string_view<CharT, Traits>::data_at(nonstd::sv_lite::basic_string_view<CharT, Traits>::size_type) const [with CharT = char; Traits = std::char_traits<char>; nonstd::sv_lite::basic_string_view<CharT, Traits>::const_reference = const char&; nonstd::sv_lite::basic_string_view<CharT, Traits>::size_type = long unsigned int]: Assertion `pos < size()' failed.

Simplified:

    a.out: simdjson.h:1357: blah blah blah: Assertion `pos < size()' failed.

On my reading of the JSON Pointer RFC 6901, the empty string is a valid JSON Pointer. Just as walking `"/foo"` from an element `e` returns a child of `e`, walking `""` from `e` should return `e` itself.

**Comment by lemire** [comment]
I will check it out in a few minutes. The `at_pointer` stuff is new and so I would not be surprised if there were issues.

**Comment by lemire** [comment]
We already test for the empty string in at_pointer. I have added a test in this PR:

https://github.com/simdjson/simdjson/pull/1148

I cannot reproduce the issue...

The following and related tests work fine...


```C++
  auto example_json = R"([1,2,{"1":"bla"}])"_padded;
  dom::parser parser;
  dom::element example = parser.parse(example_json);
  auto o = dom::array(example).at(2).at_pointer(""); // should be {"1":"bla"}
```

It works. 

I did find a bug when applying the empty string on an atomic element (e.g., `dom::array(example).at(0).at_pointer("");`  would fail). But it would not cause the error that you report, rather it would incorrectly state that the JSON Pointer had a syntax issue.

So currently, this issue is unverified. Can you provide a reproducing test case ? E.g., can you fill in the blanks in the following...

```C++
  auto example_json = R"([insert your string here)"_padded;
  dom::parser parser;
  dom::element example = parser.parse(example_json).at_pointer("");
```

I just need the content of your JSON document.

> On my reading of the JSON Pointer RFC 6901, the empty string is a valid JSON Pointer. Just as walking "/foo" from an element e returns a child of e, walking "" from e should return e itself.

That is my interpretation as well. And we do test for this very case specifically in the code. When the empty string is detected, then the current element is returned.

**Comment by nigeltao** [comment]
Take the program from https://nigeltao.github.io/blog/2020/jsonptr.html#filtering-during-not-after-parsing

Look for this line

    simdjson::dom::element element = parser.parse(input_array, input_length);

Immediately following that, replace

    if (have_query && !query.empty()) {
      element = element.at_pointer(query);
    }

with

    element = element.at_pointer("");

Compile and run on either a number or an empty JSON array

    $ g++ main.cpp simdjson.cpp -o my-simdjson
    $ echo '1' | ./my-simdjson 
    terminate called after throwing an instance of 'simdjson::simdjson_error'
      what():  Invalid JSON pointer syntax.
    Aborted (core dumped)
    $ echo '[]' | ./my-simdjson 
    my-simdjson: simdjson.h:1357: constexpr const CharT& nonstd::sv_lite::basic_string_view<CharT, Traits>::data_at(nonstd::sv_lite::basic_string_view<CharT, Traits>::size_type) const [with CharT = char; Traits = std::char_traits<char>; nonstd::sv_lite::basic_string_view<CharT, Traits>::const_reference = const char&; nonstd::sv_lite::basic_string_view<CharT, Traits>::size_type = long unsigned int]: Assertion `pos < size()' failed.
    Aborted (core dumped)

Ubuntu 18.04 LTS, g++ 7.5.

**Comment by nigeltao** [comment]
On the `1` input, it's `Invalid JSON pointer syntax`, but on the `[]` input, it's `blah blah blah: Assertion pos < size() failed`.

**Comment by nigeltao** [comment]
This fixes the [] input. The {} input obviously needs a similar fix. Not sure if there's any other places to fix, or if it's the right fix.

```diff
diff --git a/include/simdjson/dom/array-inl.h b/include/simdjson/dom/array-inl.h
index 267cc477..f0ff1c08 100644
--- a/include/simdjson/dom/array-inl.h
+++ b/include/simdjson/dom/array-inl.h
@@ -62,12 +62,10 @@ inline size_t array::size() const noexcept {
   return tape.scope_count();
 }
 inline simdjson_result<element> array::at_pointer(std::string_view json_pointer) const noexcept {
-  if(json_pointer[0] != '/') {
-    if(json_pointer.empty()) { // an empty string means that we return the current node
-      return element(this->tape); // copy the current node
-    } else { // otherwise there is an error
-      return INVALID_JSON_POINTER;
-    }
+  if(json_pointer.empty()) { // an empty string means that we return the current node
+    return element(this->tape); // copy the current node
+  } else if(json_pointer[0] != '/') { // otherwise there is an error
+    return INVALID_JSON_POINTER;
   }
   json_pointer = json_pointer.substr(1);
   // - means "the append position" or "the element after the end of the array"
```

**Comment by lemire** [comment]
Thank you.

**Comment by lemire** [comment]
This is new code that I wrote hastily to patch up our standard compliance. It is obviously unsafe.

**Comment by lemire** [comment]
This is an actual bug, evidently.

**Comment by lemire** [comment]
It is annoying that our tests did not flag this issue.

**Comment by lemire** [comment]
It is damn annoying... see this repo...

https://github.com/lemire/tao_bug_repro/blob/master/README.md

It passes even with the sanitizers.

@nigeltao  How do you trigger the warnings?

**Comment by nigeltao** [comment]
> How do you trigger the warnings?

I'm not sure what to say other than I don't use any compiler flags (other than `-o`, obviously) in

    $ g++ main.cpp simdjson.cpp -o my-simdjson

Also: Ubuntu 18.04 LTS, g++ 7.5. I haven't tried other compilers or versions.

**Comment by lemire** [comment]
Oh. I think I know. When you pass the empty string generated from "", it has actually a size of 1, so the code is safe. Let me see if I can get it to fail.

**Comment by lemire** [comment]
Fixed in our main branch.

Thanks for the report.

**Comment by pauldreik** [comment]
Just verified that the fuzzer proposed in https://github.com/simdjson/simdjson/pull/1163 detects this. The current master works fine, but if I revert the fix in 0552335ec1861f1e3094167815bc63a499b6800d a segfault happens.

**Comment by lemire** [comment]
@pauldreik Fantastic. That's the kind of testing we need!
